samza-core/src/main/scala/org/apache/samza/diagnostics/DiagnosticsManager.java

```diff
@@ -0,0 +1,173 @@
+import java.time.Duration;
+import java.time.Instant;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.concurrent.Executors;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.TimeUnit;
+import org.apache.samza.metrics.reporter.Metrics;
+import org.apache.samza.metrics.reporter.MetricsHeader;
+import org.apache.samza.metrics.reporter.MetricsSnapshot;
+import org.apache.samza.runtime.LocalContainerRunner;
+import org.apache.samza.serializers.MetricsSnapshotSerdeV2;
+import org.apache.samza.system.OutgoingMessageEnvelope;
+import org.apache.samza.system.SystemProducer;
+import org.apache.samza.system.SystemStream;
+import org.apache.samza.util.Util;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import scala.Tuple2;
+import scala.collection.JavaConverters;
+
+
+/**
+ * Responsible for publishing data to the diagnostic stream.
+ * Currently emits exception/error events obtained using a customer-appender that attaches to the root-logger.
+ */
+public class DiagnosticsManager {
+
+  private static final Logger LOG = LoggerFactory.getLogger(DiagnosticsManager.class);
+  private static final Duration DEFAULT_PUBLISH_PERIOD = Duration.ofSeconds(60);
+  // Period size for pushing data to the diagnostic stream
+
+  private static final String PUBLISH_THREAD_NAME = "DiagnosticsManager Thread-%d";
+  private static final String METRICS_GROUP_NAME = "org.apache.samza.container.SamzaContainerMetrics";
+  // Using SamzaContainerMetrics as the group name to maintain compatibility with existing diagnostics
+
+  // Parameters used for populating the MetricHeader when sending diagnostic-stream messages
+  private final String jobName;
+  private final String jobId;
+  private final String containerName;
+  private final String executionEnvContainerId;
+  private final String taskClassVersion;
+  private final String samzaVersion;
+  private final String hostname;
+  private final Instant resetTime;
+
+  private SystemProducer systemProducer; // SystemProducer for writing diagnostics data
+  private BoundedList<DiagnosticsExceptionEvent> exceptions; // A BoundedList for storing DiagnosticExceptionEvent
+  private final ScheduledExecutorService scheduler; // Scheduler for pushing data to the diagnostic stream
+  private final SystemStream diagnosticSystemStream;
+
+  public DiagnosticsManager(String jobName, String jobId, String containerName, String executionEnvContainerId,
+      String taskClassVersion, String samzaVersion, String hostname, SystemStream diagnosticSystemStream,
+      SystemProducer systemProducer) {
+    this.jobName = jobName;
+    this.jobId = jobId;
+    this.containerName = containerName;
+    this.executionEnvContainerId = executionEnvContainerId;
+    this.taskClassVersion = taskClassVersion;
+    this.samzaVersion = samzaVersion;
+    this.hostname = hostname;
+    resetTime = Instant.now();
```

Comment:
Is this still relevant? Maybe just do an Instant.now() on line 153?

Reply:
Yes in MetricsHeader reset-time and time are different things. 
Reset-time = container startup timestamp. 
time = timestamp at time of emitting metric.